tempesta_fw/http2/hindex.c

```diff
@@ -0,0 +1,882 @@
+		}
+		return 0;
+	} else {
+		DPRINTF("Unknown index: %u\n", index);
+		return Err_HPack_UnknownIndex;
+	}
+}
+
+HTTP2Index *
+hpack_new_index(unsigned int window, unsigned char is_encoder,
+		TfwPool * __restrict pool)
+{
+	HTTP2Index *const __restrict ip =
+	    tfw_pool_alloc(pool, sizeof(HTTP2Index));
+	if (ip) {
+		Sub *__restrict sub;
+
+		ip->window = window;
+		ip->size = 0;
+		ip->n = 0;
+		ip->current = 0;
+		ip->length = 0;
+		ip->entries = NULL;
+		ip->pool = pool;
+		sub = Sub_New("hpack strings", sizeof(HPackStr), 32, 32, pool);
+		if (unlikely(sub == NULL)) {
+ B0:
+			tfw_pool_free(pool, ip, sizeof(HTTP2Index));
+			return NULL;
+		}
+		ip->sub = sub;
+		ip->is_encoder = is_encoder;
+		if (is_encoder) {
+			Hash *const __restrict ht =
+			    Hash_New("hpack-hash", 0, 32, 32,
+				     hpack_hash,
+				     hpack_equal, pool);
+			Hash *__restrict hp;
+			Hash *__restrict hn;
+
+			if (unlikely(ht == NULL)) {
+ B1:
+				Sub_Delete(sub);
+				goto B0;
+			}
+			hp = Hash_New("hpack-pairs", 0, 32, 32,
+				      hpack_pair_hash, hpack_pair_equal, pool);
+			if (unlikely(hp == NULL)) {
+ B2:
+				Hash_Free(ht);
+				goto B1;
+			}
+			hn = Hash_New("hpack-names", 0, 32, 32,
+				      hpack_name_hash, hpack_name_equal, pool);
+			if (hn) {
+				ip->hash = ht;
+				ip->pairs = hp;
+				ip->names = hn;
+			} else {
+				Hash_Free(hp);
+				goto B2;
+			}
+		}
+	}
+	return ip;
+}
+
+void
+hpack_free_index(HTTP2Index * __restrict ip)
+{
+	HPackEntry *const __restrict entries = ip->entries;
+	TfwPool *const __restrict pool = ip->pool;
+
+	if (entries) {
+		tfw_pool_free(pool, entries, ip->length * sizeof(HPackEntry));
+	}
+	if (ip->is_encoder) {
+		Hash_Free(ip->names);
+		Hash_Free(ip->pairs);
+		Hash_Free(ip->hash);
+	}
+	Sub_Delete(ip->sub);
+	tfw_pool_free(pool, ip, sizeof(HTTP2Index));
+}
+
+static TfwPool *__restrict global_pool;
+
+void
+hpack_index_init(TfwPool * __restrict pool)
+{
+	unsigned int i;
+	Hash *const __restrict ht = Hash_New("static",
+					     HPACK_STATIC_ENTRIES, 0, 16,
+					     hpack_hash,
+					     hpack_equal, pool);
+	Hash *const __restrict hp = Hash_New("static-pairs",
+					     HPACK_STATIC_ENTRIES, 0, 0,
+					     hpack_pair_hash,
+					     hpack_pair_equal, pool);
+	Hash *const __restrict hn = Hash_New("static-names",
+					     HPACK_STATIC_ENTRIES, 0, 0,
+					     hpack_pair_hash,
+					     hpack_pair_equal, pool);
+
+	global_pool = pool;
+	static_hash = ht;
+	static_pairs = hp;
+	static_names = hn;
+	for (i = 0; i < HPACK_STATIC_ENTRIES; i++) {
+		HPackStr *__restrict np = (HPackStr *) & static_data[i].name;
+		HPackStr *__restrict vp = (HPackStr *) & static_data[i].value;
+
+		if (vp->len == 0) {
+			vp = NULL;
+		}
+		static_table[i].name = np;
+		static_table[i].value = vp;
+		np = hpack_hash_add(ht, np, NULL);
+		if (vp) {
+			vp = hpack_hash_add(ht, vp, NULL);
+		}
+		Hash_Add(hp, static_table + i, static_table + i);
+		Hash_Add(hn, np, static_table + i);
```

Comment:
Agree that approach with static hash table is overcomplicated. But with binary search tree - disturb the necessity of full string comparisons on each node. I'll think deeper about proposed structure. Also some kind of prefix tree may be a suitable variant here (will think in this direction too).

Reply:
Agree that approach with static hash table is overcomplicated. But with binary search tree - disturb the necessity of full string comparisons on each node. I'll think deeper about proposed structure. Also some kind of prefix tree may be a suitable variant here (will think in this direction too).

Reply:
I came to the idea exactly from patricia tree. The point is that in the example above we can, and probably, should treat the stings as long integers and compare longs instead of strings. We can set the most significant bit for the pointers and string lengths to guarantee that if we try to match 8-byte integer with say 3-byte string the set most significant bit in 4th byte will make the difference. Next idea is that we have 32-byte SIMD registers and, if we place the strings in the array with known offsets, we can load several string prefixes and match them concurrently. I have no read to implement concept, but there are several opportunities to couple fast search and efficient insert and eviction.